library/ic-certified-map/src/rbtree.rs

```diff
@@ -888,6 +888,95 @@ impl<K: 'static + AsRef<[u8]>, V: AsHashTree + 'static> RbTree<K, V> {
     }
 }
 
+use candid::CandidType;
+
+impl<'t, K, V> CandidType for RbTree<K, V> 
+where
+    K: CandidType + AsRef<[u8]> + 't,
+    V: CandidType + AsHashTree + 't,
+{
+    fn _ty() -> candid::types::internal::Type {
+        <Vec<(&K, &V)> as CandidType>::_ty()
+    }
+    fn idl_serialize<S: candid::types::Serializer>(&self, serializer: S) -> Result<(), S::Error> {
+        let collect_as_vec = self.iter().collect::<Vec<(&K, &V)>>();
+        <Vec<(&K, &V)> as CandidType>::idl_serialize(&collect_as_vec, serializer)
+    }
+}
+
+use serde::{
+    de::{Deserialize, Deserializer, MapAccess, Visitor},
+    ser::{Serialize, SerializeMap, Serializer},
+};
+use std::marker::PhantomData;
+
+impl<'t, K, V> Serialize for RbTree<K, V>
+where
+    K: Serialize + AsRef<[u8]> + 't,
+    V: Serialize + AsHashTree + 't,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let collect_refs = self.iter().collect::<Vec<(&K, &V)>>();
```

Comment:
Why is the iterator being buffered into a vector instead of serialized directly?

Reply:
Some binary formats like bincode need to know the length of the map upfront and that gets passed to the `serialize_map` function on the next line. The comments here say more: https://docs.rs/serde/latest/serde/ser/trait.Serializer.html#tymethod.serialize_map.

Reply:
Now in commit https://github.com/dfinity/cdk-rs/pull/399/commits/a1779b4adaf31db6c85aed6090af72b58797aabe it uses self.iter().count() instead of allocating a vector.